frontend_tests/node_tests/compose_ui.js

```diff
@@ -229,3 +238,51 @@ run_test("compute_placeholder_text", () => {
     opts.private_message_recipient = "alice@zulip.com,bob@zulip.com";
     assert.equal(compose_ui.compute_placeholder_text(opts), i18n.t("Message Alice, Bob"));
 });
+
+run_test("quote_and_reply", (override) => {
+    const selected_message = {
+        type: "stream",
+        stream: "devel",
+        topic: "python",
+        sender_full_name: "Steve Stephenson",
+        sender_id: 90,
+    };
+    hash_util.by_conversation_and_time_uri = () =>
+        "https://chat.zulip.org/#narrow/stream/92-learning/topic/Tornado";
+    override(current_msg_list, "selected_message", () => selected_message);
+    override(current_msg_list, "selected_id", () => 100);
+
+    let success_function;
+    override(channel, "get", (opts) => {
+        success_function = opts.success;
+    });
+
+    $("#compose-textarea").caret = make_textbox_caret_function($("#compose-textarea"));
+
+    function set_compose_content_with_caret(content) {
+        const caret_position = content.indexOf("|");
+        content = content.slice(0, caret_position) + content.slice(caret_position + 1); // remove the "|"
+        $("#compose-textarea").val(content);
+        $("#compose-textarea").caret(caret_position);
+        $("#compose-textarea").trigger("focus");
+    }
+    function get_compose_content_with_caret() {
+        const caret_position = $("#compose-textarea").caret();
+        const content =
+            $("#compose-textarea").val().slice(0, caret_position) +
+            "|" +
+            $("#compose-textarea").val().slice(caret_position); // insert the "|"
+        return content;
+    }
+
+    set_compose_content_with_caret("hello |there"); // "|" is used to encode/display position of focus before change
+    compose_actions.quote_and_reply();
+    assert.equal(get_compose_content_with_caret(), "[Quoting…]\n|hello there");
+    success_function({
+        raw_content: "Testing caret position",
+    });
+    assert.equal(
+        get_compose_content_with_caret(),
+        "translated:| @_**Steve Stephenson|90** [said](https://chat.zulip.org/#narrow/stream/92-learning/topic/Tornado):\n```quote\nTesting caret position\n```\nhello there",
```

Comment:
I think your bug is probably that `compose_ui.replace_syntax` is unexpectedly not moving the cursor?  Not sure. 

(`translated:` has a similar length to `[Quoting...]`)

Reply:
> (`translated:` has a similar length to `[Quoting...]`)

This is definitely relevant because changing `[Quoting...]` to some other string moves the position of the cursor we get.

I'm strongly suspicious that it's just jquery `$(...).val("...")` that moves the cursor to the end of the inserted text (which is just the end of the input because it completely replaces existing text) and our zjquery stub for `.val()` doesn't replicate the behavior (because it doesn't replicate any cursor behavior).

In contrast, `$(...).caret("...")` only moves the cursor forward a set amount.

Reply:
Instead of using the `replace_syntax` function, what if we use the [jquery.caret](https://github.com/acdvorak/jquery.caret)'s `range` method? Both `replace_syntax` and `range` does not give us the undo feature, but with `range`, the caret position will automatically be set at the right position and we won't need to set it ourselves. Also, it would allow us to write tests for it without making any potential changes to zjquery stub for `.val()`.

Reply:
Oh, I just realized that we are planning to get rid of the `jquery-caret` library completely in #17855.